optimizer: some cosmetic improvements on inlining

aviatesk opened the pull request:
- remove an unnecessary field and allocation
- make the signature order consistent across the file: 
  `ir::IRCode, idx::Int, stmt, rt, info, flag::UInt8, sig::Signature, state::InliningState, todo::Vector{Pair{Int, Any}}`
- pass the same `stmt::Expr` object and slightly improve its type stability

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -925,8 +923,9 @@ function call_sig(ir::IRCode, stmt::Expr)
+function inline_apply!(
+    ir::IRCode, idx::Int, stmt::Expr, sig::Signature,
+    state::InliningState, todo::Vector{Pair{Int, Any}})
     stmt = ir.stmts[idx][:inst]
```
Though I think we should move towards passing around `ir.stmt[idx]` objects, instead of unpacking them early everywhere
```suggestion
```

aviatesk replied:
@vtjnash Do you think fc033c008b3324dcf47aa2b98c904187f37d7836 is a good refactoring?

vtjnash replied:
I don't really like having an expensive and tedious copy operation instead of using a lightweight Instruction directly to access the information directly, when possible

aviatesk replied:
Hm, rather I thought having `InstInfo` passed around would reduce the cost of array indexing that otherwise happens at many places to unpack information. But creating `InstInfo`s would defeat this purpose by introducing high overheads involved with those allocations?

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1120,13 +1119,14 @@ function process_simple!(ir::IRCode, todo::Vector{Pair{Int, Any}}, idx::Int, sta
     end
 
-    return sig
+    return stmt, sig
```
this tuple looks very unfortunate

aviatesk replied:
I believe this tuple is okay since `sig` is a `Signature` object and so the tuple type is just concrete.

oscardssmith commented on base/compiler/ssair/inlining.jl:
```diff
@@ -46,6 +63,11 @@ struct ConstantCase
 
+struct SomeCase
+    val::Any
+    SomeCase(val) = new(val)
```
isn't this constructor useless?

aviatesk replied:
There is the module-wide `@nospecialize` annotation, and so this constructor is equivalent to `SomeCase(@nospecialize val) = new(val)`. Maybe the equivalent constructor would be inserted without this though.

oscardssmith commented:
now that this isn't purely cosmetic, is it worth doing some timings to make sure this isn't a compiler performance regression?

aviatesk committed:
optimizer: some cosmetic improvements on inlining

- remove an unnecessary field and allocation
- make the signature order consistent across the file:
  `ir::IRCode, idx::Int, stmt, rt, info, flag::UInt8, sig::Signature, state::InliningState, todo::Vector{Pair{Int, Any}}`
- pass the same `stmt::Expr` object and slightly improve its type-stability

aviatesk commented:
Alright, I removed `InstInfo` stuff and I'd like to move this forward with the original "cosmetic" changes.
I will merge this once CI goes green.
